### PR TITLE
refactor: Make ToxAV independent of toxcore internals.

### PR DESCRIPTION
**Pull request by iphydf**
closes #1431

This change is [Reviewable](https://reviewable.io/reviews/TokTok/c-toxcore/2651)

### CODE REVIEW

**Comment by iphydf** [comment]
This is ready for review.

File: toxcore/tox.c [hunk 1]
```diff
@@ -1072,6 +1072,16 @@ void tox_kill(Tox *tox)
     mem_delete(tox->sys.mem, tox);
 }
 
+void tox_get_options(Tox *tox, struct Tox_Options *options)
```
Comment (Green-Sky) [inline toxcore/tox.c:1075]: what is this doing in this pr? Is this some left over?

File: toxav/toxav.c [hunk 2]
```diff
@@ -193,6 +239,9 @@ ToxAV *toxav_new(Tox *tox, Toxav_Err_New *error)
     init_decode_time_stats(&av->video_stats);
     av->msi->av = av;
 
+    // save Tox object into toxcore
```
Comment (Green-Sky) [inline toxav/toxav.c:242]: *ToxAV object

File: toxcore/tox.c (hunk 1 above)
Comment (Green-Sky) [reply toxcore/tox.c]: I extracted the code here https://gist.github.com/Green-Sky/60199323907fdc3862196da938abdedf

**Review by Green-Sky — CHANGES_REQUESTED** [review]
`toxav_old.c` uses the `Messenger` logger, instead the `ToxAV` logger.

File: toxav/audio.h [hunk 3]
```diff
@@ -61,9 +66,10 @@ typedef struct ACSession {
     /* Audio frame receive callback */
     toxav_audio_receive_frame_cb *acb;
     void *acb_user_data;
+    Tox *tox;
 } ACSession;
 
-ACSession *ac_new(Mono_Time *mono_time, const Logger *log, ToxAV *av, uint32_t friend_number,
+ACSession *ac_new(Mono_Time *mono_time, const Logger *log, Tox *tox, ToxAV *av, uint32_t friend_number,
```
[the remark this answers is not in the document]
Comment (Green-Sky) [reply toxav/audio.h:72]: (I am referring to to the `tox` field in `ACSession` a couple of lines up.

File: toxav/msi.c [hunk 4]
```diff
@@ -9,12 +9,22 @@
...
+#include "../toxcore/tox.h"
+#include "../toxcore/tox_private.h"
 #include "../toxcore/util.h"
 
 #define MSI_MAXMSG_SIZE 256
 
+#ifndef TOXAV_DEFINED
+#define TOXAV_DEFINED
+typedef struct ToxAV ToxAV;
+#endif /* TOXAV_DEFINED */
```
Comment (Green-Sky) [inline toxav/msi.c:26]: we are including `toxav_hacks.h` which provides functions that already need this, so it is safe to assume this already exists here.

File: toxav/msi.c [hunk 5]
```diff
@@ -163,21 +173,57 @@ int msi_kill(MSISession *session, const Logger *log)
     free(session);
     return 0;
 }
-int msi_invite(MSISession *session, MSICall **call, uint32_t friend_number, uint8_t capabilities)
+
+/*
+ * return true if friend was offline and the call was canceled
```
Comment (Green-Sky) [inline toxav/msi.c:178]: is offline.

File: toxav/bwcontroller.h [hunk 6]
```diff
@@ -5,19 +5,24 @@
 #ifndef C_TOXCORE_TOXAV_BWCONTROLLER_H
 #define C_TOXCORE_TOXAV_BWCONTROLLER_H
 
-#include "../toxcore/Messenger.h"
+#include <stdint.h>
```
Comment (Green-Sky) [inline toxav/bwcontroller.h:8]: standard headers always after your own headers.

File: toxcore/tox.h [hunk 7]
```diff
@@ -3288,6 +3288,9 @@ bool tox_friend_send_lossless_packet(
     Tox_Err_Friend_Custom_Packet *error);
 
 /**
+ * tox_callback_friend_lossy_packet is the compatibility function to
+ * set callback for all packet IDs except those reserved for ToxAV
```
Comment (Green-Sky) [inline toxcore/tox.h:3292]: .

File: toxcore/Messenger.h [hunk 8]
```diff
@@ -200,15 +200,11 @@ typedef void m_friend_lossy_packet_cb(Messenger *m, uint32_t friend_number, uint
...
-typedef void m_friend_connectionstatuschange_internal_cb(Messenger *m, uint32_t friend_number,
-        bool is_online, void *user_data);
 typedef void m_conference_invite_cb(Messenger *m, uint32_t friend_number, const uint8_t *cookie, uint16_t length,
                                     void *user_data);
 typedef void m_group_invite_cb(const Messenger *m, uint32_t friend_number, const uint8_t *invite_data, size_t length,
                                const uint8_t *group_name, size_t group_name_length, void *user_data);
-typedef void m_msi_packet_cb(Messenger *m, uint32_t friend_number, const uint8_t *data, uint16_t length,
-                             void *user_data);
-typedef int m_lossy_rtp_packet_cb(Messenger *m, uint32_t friend_number, const uint8_t *data, uint16_t length, void *object);
+typedef int m_lossy_rtp_packet_cb(Messenger *m, uint32_t friend_number, const uint8_t *data, uint16_t len, void *object);
```
Comment (Green-Sky) [inline toxcore/Messenger.h:207]: len? -> length

**Review by Green-Sky — CHANGES_REQUESTED** [review]

File: toxcore/net_crypto.c [hunk 9]
```diff
@@ -1863,30 +1818,14 @@ static int create_crypto_connection(Net_Crypto *c)
...
-        }
-
         // Memsetting float/double to 0 is non-portable, so we explicitly set them to 0
         c->crypto_connections[id].packet_recv_rate = 0.0;
         c->crypto_connections[id].packet_send_rate = 0.0;
         c->crypto_connections[id].last_packets_left_rem = 0.0;
         c->crypto_connections[id].packet_send_rate_requested = 0.0;
         c->crypto_connections[id].last_packets_left_requested_rem = 0.0;
-        c->crypto_connections[id].mutex = mutex;
-        c->crypto_connections[id].status = CRYPTO_CONN_NO_CONNECTION;
```
Comment (goldroom) [inline toxcore/net_crypto.c:1886]: Is this intentional? This is the only place where `crypto_connection` is set to this state.

Comment (Green-Sky) [reply toxcore/net_crypto.c:1886]: `#define CRYPTO_CONN_NO_CONNECTION 0`
It used to be defined as the zero.

Comment (Green-Sky) [reply toxcore/net_crypto.c:1886]: The mutex is very intentional, status I will check.

Comment (zoff99) [reply toxcore/net_crypto.c:1886]: let me check. this could be a mistake. since this line is there in my fork and in toktok master ...

Comment (zoff99) [reply toxcore/net_crypto.c:1886]: ok i see what happend here:
first the commit 6338cb46ad99dbc2806976aefabbf2e98476ebda
refactored the function to put c->crypto_connections[id].status = CRYPTO_CONN_NO_CONNECTION; in a different place.
then after that commit 4db6cfb19239a0d8b83ef91b1cc26fab1979fca7 most likely removed the line by mistake in a merge conflict resolve

Comment (Green-Sky) [reply toxcore/net_crypto.c:1886]: After looking at the code, I think we dont need that enum value.
`CRYPTO_CONN_FREE` represents the same thing. But now we need to make sure, we don't improperly reuse the connection without memsetting it...

Comment (zoff99) [reply toxcore/net_crypto.c:1886]: lets put this line back please

Comment (Green-Sky) [reply toxcore/net_crypto.c:1886]: ah nice, github did not update @zoff99 in my open page. Yes, lets add the line back in and figure out if we need `CONN_NO_CONNECTION` at all somewhere else.

File: toxcore/Messenger.h (hunk 8 above)
Comment (zoff99) [reply toxcore/Messenger.h:207]: good catch.
"m_lossy_rtp_packet_cb" is actually a leftover that should be deleted. same as "RTP_Packet_Handler"

Comment (zoff99) [reply toxcore/Messenger.h:207]: do we have some sort of "identify unused functions and struct defs" scanner?

**Comment by zoff99** [comment]
@Green-Sky "m_lossy_rtp_packet_cb" is actually a leftover that should be deleted. same as "RTP_Packet_Handler"

**Comment by Green-Sky** [comment]
> @Green-Sky "m_lossy_rtp_packet_cb" is actually a leftover that should be deleted. same as "RTP_Packet_Handler"

True, fixed.

File: toxcore/Messenger.h (hunk 8 above)
Comment (zoff99) [reply toxcore/Messenger.h]: removed

File: toxav/bwcontroller.h (hunk 6 above)
Comment (zoff99) [reply toxav/bwcontroller.h:8]: ok so if its everywhere then we can not solve it in this PR. its a different topic then

File: toxav/rtp.c [hunk 10]
```diff
@@ -647,36 +674,44 @@ size_t rtp_header_unpack(const uint8_t *data, struct RTPHeader *header)
...
+        LOGGER_ERROR(log, "out of memory while allocating work buffer list");
         free(session);
         return nullptr;
     }
 
     // First entry is free.
     session->work_buffer_list->next_free_entry = 0;
 
-    session->ssrc = payload_type == RTP_TYPE_VIDEO ? 0 : random_u32(m->rng);
+    session->ssrc = payload_type == RTP_TYPE_VIDEO ? 0 : rtp_random_u32(); // Zoff: what is this??
```
[the remark this answers is not in the document]
Comment (zoff99) [reply toxav/rtp.c:706]: we are not using RTP. we will remove that random value and set it to zero in another PR, after merging this one.
do you want to open an issue for that?

**Comment by zoff99** [comment]
@Green-Sky mergeable is failing because of the commit messages missing "refactor" or similar

**Comment by Green-Sky** [comment]
> @Green-Sky mergeable is failing because of the commit messages missing "refactor" or similar

Yes, that is going to go away when I squash this pr.

File: toxav/bwcontroller.c [hunk 11]
```diff
@@ -190,28 +192,41 @@ static int on_update(BWController *bwc, const struct BWCMessage *msg)
...
-}
-
-static int bwc_handle_data(Messenger *m, uint32_t friend_number, const uint8_t *data, uint16_t length, void *object)
-{
-    BWController *bwc = (BWController *)object;
+    const Logger *log = toxav_get_logger(toxav);
 
     if (length - 1 != sizeof(struct BWCMessage)) {
-        return -1;
+        LOGGER_ERROR(log, "Actual data size and length argument do not match");
```
[the remark this answers is not in the document]
Comment (Green-Sky) [reply toxav/bwcontroller.c]: https://github.com/TokTok/c-toxcore/issues/2781

**Review by Green-Sky — APPROVED** [review]

**Review by iphydf — COMMENTED** [review]
> Reviewed 12 of 26 files at r1, 3 of 4 files at r2, 3 of 5 files at r3, 8 of 8 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/toktok/c-toxcore/2651)* status: 0 of 1 approvals obtained (waiting on @goldroom and @Green-Sky)

**Review by Green-Sky — APPROVED** [review]
> Reviewed 12 of 26 files at r1, 3 of 4 files at r2, 3 of 5 files at r3, 8 of 8 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/toktok/c-toxcore/2651)* status: 0 of 1 approvals obtained (waiting on @goldroom)

**Pull request merged**